[Vending machine step3] MVC 패턴

1Consumption opened the pull request:
중복되는 IBOutlet을 없애기 위해 IBOutletCollecton을 사용하여 각 view의 태그값으로 접근하도록 하였습니다. 
beverage 프로퍼티를 enum으로 설정하여 값을 제한함과 동시에 명확히 하였습니다.
view의 변화를 감지하여 controller가 model의 값을 바꾸면 Notification을 사용하여 controller가 다시 view를 변경하도록 하였습니다. 

[아쉬운 점]
스토리보드로 뷰를 설정하려니 생각보다 어려워서 제대로 못한 것 같습니다. 이 부분에 대해서는 더 공부해야된다고 생각이 들었습니다.

1Consumption committed:
[기능 수정] 매개변수 추가

보다 유연한 메소드를 위해 기준이 되는 매개변수 추가

1Consumption committed:
[기능 수정] testInsertBeverage() 메소드

해당 메소드가 기능을 잘 하는지 확인하기 위해 cider 객체를 insert 함

1Consumption committed:
[파일 추가] 음료 이미지

1Consumption committed:
[기능 추가] 자판기 UI 설정

1Consumption committed:
[기능 추가] stroyBoard와 viewController 연결
IBOutletCollcetion을 사용하여 버튼들과 이미지뷰, 라벨을 효율적으로 관리

1Consumption committed:
[기능 수정] view 배경 변경
view 배경 색 변경 및 stackview spacing 조정
imageView content mode: Aspect fit -> Aspect Fill로 변경

1Consumption committed:
[기능 추가] imageView 테두리 둥글게 변경

1Consumption committed:
[기능 추가] 잔액 표시

1Consumption committed:
[기능 수정] Beverage 프로퍼티 수정
Beverage 프로퍼티의 속성을 Enum으로 변경함
Beverage 객체를 제작할 때, 여러가지 조건을 조합해서 제작하는 것이 아닌, "추가" 버튼을 누르면 추가가 되는 방식이기 때문에 수정 함. 또한 자판기에 들어가는 음료의 종류는 한정적이기 때문에 enum으로 값을 제한 해줌.
후에 Factoty클래스를 만들어서 Beverage 객체를 만들 예정임.

1Consumption committed:
[기능 추가] BeverageFactory 클래스 추가
view에서 선택한 "추가" 버튼의 태그에 따라 음료를 추가하기 위해 BeverageFactory 클래스를 만들어줌
BeverageFactory 클래스에서 makeBeverage 메소드가 전달된 음료 번호에 맞게 Beverage 객체를 리턴해줌.

또한 VendingMachine 클래스에서 Beverage 객체를 추가할 때, 버튼의 태그를 매개변수로 받게 수정 함.

1Consumption committed:
[기능 추가] 최소 stockLabel 설정

1Consumption committed:
[기능 추가] 재고 추가 기능 완료

음료별로 고유 인덱스를 부여하기 위해 각 클래스에 음료 번호 프로퍼티를 추가해 줌
Notification을 활용하여 재고의 변동을 알아챔

1Consumption committed:
[기능 추가] 잔액 표시 기능 구현

Notification을 통해 잔액이 변경됨을 알아챔.
버튼의 tag만큼 잔액을 더해줌

1Consumption committed:
[Readme 수정] Readme 수정

스크린샷 첨부

godrm commented on VendingMachineApp/VendingMachineApp/Model/Beverage/Beverage.swift:
```diff
@@ -9,47 +9,98 @@
+        case Coke = "coke"
+        case Cider = "cider"
+        case GeorGia = "georgia"
+        case TOP = "TOP"
```
브랜드나 이름은 enum 타입으로 하는 건 좋은 선택이 아닌 것 같습니다.
왜냐하면 새로운 브랜드나 이름이 추가될 때마다 코드를 다시 컴파일하는 구조를 만들지 않는 게 좋기 때문입니다. 
자주 쓰는 타입을 코드로 선언해 놓는 것과 달리, 매번 다시 빌드해서 배포해야 한다면 좋은 선택이 아닙니다. 클래스 속성을 만들 때는 항상 변화에 대해 고민해보세요.

godrm commented on VendingMachineApp/VendingMachineApp/Model/Beverage/Beverage.swift:
```diff
@@ -9,47 +9,98 @@
+        
+        func isCheaper(than balance:Int) -> Bool{
+            return balance >= self.rawValue
+        }
```
Price 대신에 Money 수준으로 객체를 설계해보세요. 
음료수 뿐만 아니라 자판기에서도 사용할 수 있을 겁니다.

godrm commented on VendingMachineApp/VendingMachineApp/Model/Beverage/Beverage.swift:
```diff
@@ -9,47 +9,98 @@
+        
+        func isZero() -> Bool{
+            return self == Beverage.Calorie.Low
+        }
```
이런 양을 표현하는 타입을 선언할 때 기준을 정하는 것은 좋은데, 
넣을 수 있는 값을 고정해버리면 수정하기 어려워집니다. 
Double 타입으로 양을 표현하는 것도 나쁘지는 않습니다. isZero()처럼 isLow() isHigh() 같은 메소드가 있는 건 어떨까요? 그럴려면 기준값이 필요하기는 합니다.

godrm commented on VendingMachineApp/VendingMachineApp/Model/Beverage/Beverage.swift:
```diff
@@ -9,47 +9,98 @@
-    private let standardOfHot: Double = 65
+    private var calorie: Calorie
+    private var temperature: Temperature
+    private(set) var beverageIndex: VendingMachine.BeverageNumbers
```
음료수 데이터 값에 자판기 값을 가지는 것은 좋지 않습니다.
하위 객체가 상위 객체 타입에 의존해도 안되고, 다른 어떤 객체가 자기를 포함하는지 모르는 게 좋습니다. 
음료수 > 자판기 > 앱 수준으로 점점 상위 모듈이 되는데, 하위 모듈은 상위 모듈을 참조하거나 의존하면 안됩니다. 
index는 자판기에서만 필요한 데이터입니다.

godrm commented on VendingMachineApp/VendingMachineApp/Model/Beverage/Beverage.swift:
```diff
@@ -9,47 +9,98 @@
         self.name = name
         self.manufacturingDate = manufacturingDate
-        self.expirationDate = expirationDate
+        self.expirationDate = Date(timeInterval: 86400 * 14, since: manufacturingDate)
```
manufacturingDate 부터 며칠인지 표현하는 방법이 어떨까요?
예를 들어 `Date.init(from: manufacturingDate, afterDays: 14)` 형태는 어떨까요?

godrm commented on VendingMachineApp/VendingMachineApp/Model/Beverage/Beverage.swift:
```diff
@@ -9,47 +9,98 @@
     
     func getPrice() -> Int {
-        return price
+        return price.rawValue
```
Price라는 타입을 만들었으면 rawValue : Int 타입을 리턴하지 말고 그대로 Price 타입을 활용하세요.

godrm commented on VendingMachineApp/VendingMachineApp/Model/Beverage/Beverages.swift:
```diff
@@ -17,6 +17,8 @@ class Beverages {
     func add(beverage: Beverage) {
         beverages.append(beverage)
+        NotificationCenter.default.post(name: Notification.Name.updateStock,
+                                        object: nil, userInfo: ["index": beverage.beverageIndex.getValue(), "stock": beverages.filter{$0 == beverage}.count])
```
추가할 때는 Notification.Name.updateStock 노티가 발생하는데, remove 할 때는 필요없는건가요? 같은 맥락이지 않을까요?

godrm commented on VendingMachineApp/VendingMachineApp/Model/Beverage/Milk.swift:
```diff
@@ -9,18 +9,30 @@
+        }
+        
+        func isSame(farmCode: String) -> Bool {
+            return self.rawValue == farmCode
```
isSame(farmCode:) 보다 FarmCode가 Equatable 프로토콜을 구현하는게 어떨까요?
FarmCode라는 타입을 만들고 String을 비교할 필요없이 FarmCode 타입끼리 비교하면 되지 않을까요? 카드게임에서 카드끼리 비교했던 것처럼요

godrm commented on VendingMachineApp/VendingMachineApp/Model/Beverage/TOP.swift:
```diff
@@ -9,18 +9,22 @@
+        
+        init(index: Int) {
+            self = BeanOrigin.allCases[index]
+        }
```
모든 속성을 enum 타입으로 만들고 있는데, 속성 값이 추가될 때마다 다시 컴파일 해야하는 방식은 좋은 선택이 아닙니다. 
그리고 `init(index:Int)` 보다는 `init(with origin: BeanOrigin)` 형태로 선언한 타입을 적극 활용하세요.

godrm commented on VendingMachineApp/VendingMachineApp/Model/BeverageFacory.swift:
```diff
@@ -0,0 +1,83 @@
+                                  name: .StrawberryMilk,
+                                  manufacturingDate: Date(),
+                                  calorie: .Middle,
+                                  farmCode: Milk.FarmCode(index: Int.random(in: 0..<Milk.FarmCode.allCases.count)),
```
random하게 FarmCode를 만드는 것도 FarmCode 내부에 구현해보세요.

godrm commented on VendingMachineApp/VendingMachineApp/Model/VendingMachine.swift:
```diff
@@ -74,7 +87,7 @@ class VendingMachine{
         var hot = [String]()
         beverages.forEachBeverages {
-            if $0.isHot() {
+            if $0.isHot(standard: 65) {
```
여기서 비교하는 65 같은 기준 값을 enum으로 선언하는 것은 좋습니다.

godrm commented on VendingMachineApp/VendingMachineApp/Model/VendingMachine.swift:
```diff
@@ -37,6 +47,9 @@ class VendingMachine{
         self.balance += balance
+        NotificationCenter.default.post(name: Notification.Name.updateBalance,
+                                        object: nil,
+                                        userInfo: ["balance": self.balance])
```
잔액이 변경되었다는 노티를 add할 때만 보내고 있네요. 
아래 purchase 할 때도 보내야 하지 않을까요?

godrm commented on VendingMachineApp/VendingMachineApp/ViewController.swift:
```diff
@@ -9,20 +9,74 @@
 
 class ViewController: UIViewController {
-
+    let vendingMachine = VendingMachine()
```
ViewController에서도 내부에서만 사용하는 변수는 private let 으로 선언하세요.

godrm commented on VendingMachineApp/VendingMachineApp/ViewController.swift:
```diff
@@ -9,20 +9,74 @@
+    @IBOutlet weak var balanceLabel: UILabel!
+    
+    @IBAction func addStock(_ sender: UIButton) {
+        vendingMachine.insert(beverageNumber: VendingMachine.BeverageNumbers.allCases[sender.tag])
```
이 부분도 index만 넘기면 VendingMachine에서 판단할 수 있을까요?
참고로 tag를 쓰는 방식이 실수할 경우가 많아서 위험하기도 합니다.
기본 tag값이 0이라서 구분이 잘 안되기도 하구요.

godrm commented on VendingMachineApp/VendingMachineApp/ViewController.swift:
```diff
@@ -9,20 +9,74 @@
+    }
+    
+    @IBAction func addBalance(_ sender: UIButton) {
+        vendingMachine.add(balance: sender.tag)
```
여기서는 금액도 tag를 쓰고 있네요. 
이것보다는 tag값을 기준으로 Price 혹은 Money 객체를 만들어서 전달하는 게 좋습니다. 
그래야 오동작을 최소화할 수 있습니다.

godrm commented on VendingMachineApp/VendingMachineApp/ViewController.swift:
```diff
@@ -9,20 +9,74 @@
+            stockLabels.forEach{
+                $0.text = String(0)
+            }
+        }
```
여기 초기 뷰값을 지정하는 부분과 아래 Notification을 받아서 뷰 값을 바꾸는 부분도 일반화시켜서 동일하게 동작하도록 개선하세요.
그래야 뷰 동작을 바꿀 때도 한 곳만 바꾸면 됩니다.

godrm reviewed (commented):
enum 타입을 굉장히 적극적으로 활용했네요. 
다만 클래스 상속 관계를 다룰 때 enum 타입이 확장을 저해하는 요소가 되기도 합니다.
음료수 모델은 계속 바뀌고, 새로운 음료가 추가될 수도 있는데 
그렇게 요구사항이 변할 때 유연한 구조로 바뀌는 부분을 최소화하는 게 중요합니다.
enum 타입을 만들어놓고, 그 enum 타입을 사용하는 곳에서 비교문으로 분기해야 한다면 변화에 취약한 구조가 됩니다.
다음 피드백을 확인해보세요.

1Consumption committed:
[기능 수정] 속성 변경
brand 프로퍼티를 enum 속성인 Brand에서 String으로 변경

1Consumption committed:
[기능 수정] 속성 변경
name 프로퍼티를 enum 속성인 Name에서 String으로 변경

1Consumption committed:
[기능 수정] 속성 변경
capacity 프로퍼티를 enum 타입인 Capacity에서 Double로 변경

1Consumption committed:
[기능 추가] Price 클래스 생성 및 타입 수정
price 프로퍼티를 enum 타입인 Price에서 Price 클래스로 변경

1Consumption committed:
[기능 수정] 속성 변경
calorie 속성을 enum 타입이던 Calorie에서 Double로 변경

1Consumption committed:
[기능 수정] 속성 변경
temperature 속성을 enum 타입이던 Temperature 에서 Double로 변경

1Consumption committed:
[기능 수정] Beverage 클래스에 부여된 음료 넘버 삭제

Beverage 클래스와 VendingMachine 클래스간 의존성을 없애기 위함
이를 위해 ViewController에 selectedIndex라는 프로퍼티를 만들었음.
이 프로퍼티는 가장 최근에 눌린 버튼의 인덱스가 몇번째인지를 기억하는 프로퍼티임.
이로인해 Beverages 클래스에서는 음료 넘버를 노티피케이션할 필요 없어지고, 어떤 음료가 변경되었는지만 노티피케이션하면 됨.

1Consumption committed:
[기능 수정] 접근자 수정

1Consumption committed:
[기능 수정] 유통기한 할당 방법 수정

extension을 이용하여 가독성을 높임

1Consumption committed:
[기능 수정] 생성자 수정

인덱스를 받아 enum 타입을 생성하는 것이 아닌
생성자에서 랜덤 넘버를 발생시켜 enum 타입을 생성하도록 수정

1Consumption committed:
[기능 수정] 기준을 상수로 설정

1Consumption committed:
[기능 수정] 버튼 클릭 시 보내는 인자 변경

버튼 클릭시 버튼의 태그를 인자로 보냈으나,
버튼의 인덱스를 보내게 변경

1Consumption committed:
[기능 수정] 메소드 병합

setupBalanceLabel()과 setupStockLabels() 메소드를 각각 updateBalance(), updateStockLabel 메소드에 병합함.
또한 기존 Notification 타입을 매개변수로 받았으나, 어플 최초 실행 시를 구분하기 위해 Notification? 타입을 매개변수로 설정함.

1Consumption commented:
피드백 반영하여 수정하였습니다!

godrm commented:
코드 리뷰는 숙제 검사가 아니라, 기능 개발을 위해 고민, 설계, 과정과 결과를 다른 개발자와 소통하고 설명하고 설득하는 과정입니다. 
제가 남긴 피드백에 대해서 본인의 의견도 남겨보세요. 그래야 어떤 방향에서 어떻게 고쳤는지도 의미가 있고
왜 고쳐야 하는지도 설득하거나 설득당하는 과정을 볼 수 있습니다.

godrm commented on VendingMachineApp/VendingMachineApp/Model/Beverage/Beverage.swift:
```diff
@@ -9,92 +9,36 @@
+    private var brand: String
+    private var capacity: Double
+    private(set) var price: Price
+    private var name: String
```
기준 값으로 선언해서 쓸 수 있는 enum도 있는데 모두 지워버렸네요. 그 부분은 아쉽습니다.

1Consumption replied:
temperature 타입의 경우 뜨겁냐 차갑냐의 차이이기 때문에 다시 enum 타입으로 선언했습니다!
enum 타입의 사용시기에 대해 조금 더 알게 된 것 같습니다.

godrm commented on VendingMachineApp/VendingMachineApp/Model/Beverage/Beverage.swift:
```diff
@@ -9,92 +9,36 @@
         self.name = name
         self.manufacturingDate = manufacturingDate
-        self.expirationDate = Date(timeInterval: 86400 * 14, since: manufacturingDate)
+        self.expirationDate = Date().expiredDay(from: manufacturingDate, period: 14)
```
period가 days만을 의미한다면 days를 써주는 것도 좋을 것 같습니다.

godrm commented on VendingMachineApp/VendingMachineApp/Model/Beverage/Beverage.swift:
```diff
@@ -9,92 +9,36 @@
-    
     func isHot(standard: Double) -> Bool {
-        return temperature.isHotter(than: standard)
+        return temperature >= standard
```
enum 타입으로 여기에 들어갈 기준 값을 지정해놓는 것은 좋았을 것 같습니다.

godrm commented on VendingMachineApp/VendingMachineApp/Model/Beverage/Cider.swift:
```diff
@@ -13,16 +13,16 @@ class Cider: Soda {
-        init(index: Int) {
-            self = Taste.allCases[index]
+        init() {
+            self = Taste.allCases[Int.random(in: 0..<Taste.allCases.count)]
```
요구사항에서는 random 으로 만들라는 게 없었죠? 
매번 인스턴스를 만들때마다 다른 맛이 만들어진다면 이 객체는 어떻게 테스트할 수 있을까요?
생성 과정의 편리성을 위해서라면 객체는 매개변수로 지정할 수 있게 하고, 그걸 생성하는 팩토리 객체를 만들어서 거기서 하나를 고르도록 구현하는 게 좋습니다.

godrm commented on VendingMachineApp/VendingMachineApp/Model/BeverageFacory.swift:
```diff
@@ -9,75 +9,5 @@
-                       beanOrigin: TOP.BeanOrigin.init(index: Int.random(in: 0..<TOP.BeanOrigin.allCases.count)))
-        }
-    }
+
```
팩토리 객체가 있었네요. 팩토리 객체 역할은 무엇이었을까요?
위에 생성 시점에서 변경해야 하는 내용에 대해 참고하세요. 
필요없으면 과감하게 지워도 되는거죠.

1Consumption replied:
제가 만들었던 팩토리 객체는 단순히 매개변수에 따라 각기 다른 객체를 리턴하는 객체였습니다. 그러나 이 객체에서 가지고 있던 beverageList를 vendinMachine 객체에 넣는것이 낫겠다고 생각하여 해당 클래스를 삭제하였습니다.

godrm replied:
네. 기능이 필요없으면 과감하게 지워도 됩니다. 
다만 다른 코드에서 랜덤으로 생성해주는 기능이 있던데 그런 경우는 팩토리 객체가 하는 게 나을 것 같아요

godrm commented on VendingMachineApp/VendingMachineApp/Model/Price.swift:
```diff
@@ -0,0 +1,30 @@
+        self.money = money
+    }
+    
+    func add(moneyNumber: Int) {
```
Int가 인덱스인가 본데 100 , -100 이런게 들어오면 어떻게 되죠?
이런 경우도 Int 대신에 구체적인 타입을 정하는 게 좋습니다. 
그렇지 않다면 moneyList를 항상 오픈해서 그 중에 하나를 선택하도록 만들어야 합니다. 
항상 클래스 하나도 안전하고 독립적으로 동작하도록 설계하세요.

1Consumption replied:
구체적인 타입이라면 enum 타입으로 값을 제한하라는 말씀이신가요?

godrm commented on VendingMachineApp/VendingMachineApp/Model/VendingMachine.swift:
```diff
@@ -10,27 +10,73 @@ import Foundation
+    private(set) var balance: Price
     private var purchaseHistory: Beverages
-    private let beverageFactory: BeverageFactory
+    private let hotStandard = 65.0
```
hotStandard 는 모든 음료의 공통 사항일까요? 자판기의 상태값인가요?

godrm commented on VendingMachineApp/VendingMachineApp/Model/VendingMachine.swift:
```diff
@@ -45,28 +91,31 @@ class VendingMachine{
         var beverageList = [String]()
         beverages.forEachBeverages{
-            if $0.canBuy(have: balance) {
+            if $0.canBuy(have: balance.money) {
```
balance 타입을 만들었으면 balance를 그대로 넘겨서 처리할 수 있어야 합니다. 내부 속성을 넘기지 않는 게 좋습니다. 그래야 객체 만들어서 사용하는 게 됩니다.

godrm commented on VendingMachineApp/VendingMachineApp/Model/VendingMachine.swift:
```diff
@@ -45,28 +91,31 @@ class VendingMachine{
-        purchaseHistory.add(beverage: beverage)
-        beverages.remove(beverage: beverage)
+    func purchase(beverageNumber: Int) {
+        balance.add(money: -beverageList[beverageNumber].price.money)
```
이 부분은 어색하네요. 
balance.minus(money:) 가 더 읽기 편하지 않을까요? 
마찬가지로 price를 직접 넘기는 게 좋습니다.

godrm commented on VendingMachineApp/VendingMachineApp/ViewController.swift:
```diff
@@ -32,8 +35,8 @@ class ViewController: UIViewController {
-        setupBalanceLabel()
-        setupStockLabels()
+        updateBalance(nil)
+        updateStockLabel(nil)
```
초기값을 다루는 경우 Notification Selector를 직접 호출하기 보다는 공통 내부 함수를 만드는 게 좋습니다. 
이렇게 하면 매개변수에 따라서 분기해서 처리했을 텐데 좋은 선택이 아닙니다.

1Consumption replied:
이 부분에 대해서는 어떻게 해야할지 모르겠습니다.
notification이 와야 update를 해줄텐데 공통 내부함수를 만들라는 말이 어떤 말인지 잘 모르겠습니다.

godrm replied:
viewDidLoad() -> private 공통 함수 호출
notification_selector() -> private 공통 함수 호출 
노티를 받는 것과 초기값을 보여주는 게 결과적으로 같은 것이지만, 이렇게 하는게 낫다는 겁니다. 
아래 피드백도 함께 고려해보세요.

godrm commented on VendingMachineApp/VendingMachineApp/ViewController.swift:
```diff
@@ -66,17 +56,28 @@ class ViewController: UIViewController {
+                balanceLabel.text = "잔액: \(balance)"
+            }
+        } else {
+            balanceLabel.text = "잔액: \(vendingMachine.balance.money)"
```
이 부분도 비교문으로 구분해서 처리하지 말고 공통 함수를 만들어서 `balanceLabel.text = "잔액: \(vendingMachine.balance.money)"` 만 호출해도 되지 않나요?

godrm reviewed (commented):
세부적인 피드백을 더 남겼습니다. 이 부분에 대해서도 생각을 정리해보고 개선해보면 좋겠습니다.

1Consumption committed:
[기능 수정] hotStandard 프로퍼티 변경

VendingMachine 클래스에서 Beverage 클래스로 변경
뜨거운 여부는 음료의 속성이기 때문에 옮김.

1Consumption committed:
[기능 수정] temperature 타입 변경

temperature 프로퍼티를 enum 타입인 Temperature로 변경
이는 온도의 뜨겁고 차가움을 나타내는 기준값이 필요하기 때문

1Consumption committed:
[기능 수정] expiredDay 메소드 파라미터 명 변경

period는 구체적이지 않기 때문에 보다 구체적인 days로 변경

1Consumption committed:
[기능 수정] Cider 클래스 taste 프로퍼티 변경

랜덤으로 생성하지 않고 기본적으로 lemon을 생성하게 변경

1Consumption committed:
[기능 수정] Price 클래스 내 money 접근자 변경 및 CustomStringConvertible 프로토콜 채용

내부 변수에 접근 못하도록 private(set) 에서 private으로 수정
Price 클래스에 CustomStringConvertible프로토콜을 채용하여 money를 문자열로 출력하도록 설정

1Consumption committed:
[기능 수정] updateBalance, updateSotckLabels 메소드 변경

공통 내부함수를 사용하도록 변경

1Consumption committed:
[기능 수정] balance 인자 제한

balance를 추가할 때 값을 제한하기 위해 enum 타입을 사용함.
만일 해당 범위를 넘어설 경우 0원이 추가되도록 설정.

1Consumption commented:
구체적인 피드백 감사합니다!! 반영하여 수정하였습니다.

godrm commented on VendingMachineApp/VendingMachineApp/Model/Beverage/Beverage.swift:
```diff
@@ -9,36 +9,45 @@
+        case Hot = 65.0
+        
+        func isHot() -> Bool {
+            return self == Temperature.Hot
```
오잉? 같을 때만 hot이고 값이 99도나 66도는 안되는건가요? 
이렇게 enum 타입으로 선언한 건 비교할 때 기준 타입인지 저 값만 넣을 수 있는건지 명확하게 구분해주세요.

1Consumption replied:
그러게요... 왜 저렇게 했을까요ㅠㅠ 정말 말이 안되네요....
다음 스텝 진행하면서 고치겠습니다!
